File: contests/problemC/solutions/solution_binary_search.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> calc_max_height(int max_cows, const vector<pair<int,int>>& cows) {
    const int n = (int)cows.size();
    vector<int> count(n, 0), bound(n, 0);
    int y_bound = n - 1;
    int sum = 0;

    for (int x = 0, i = 0; x < n; ++x) {
        while (i < n && cows[i].first == x) {
            int y = cows[i].second;
            count[y]++;
            if (y <= y_bound) sum++;
            ++i;
        }
        while (sum > max_cows) {
            sum -= count[y_bound];
            --y_bound;
        }
        bound[x] = y_bound;
    }
    return bound;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    long long H, W;
    cin >> H >> W;
    int N;
    if (!(cin >> N)) return 0;

    // FIX TRÀN: x,y phải là long long
    vector<pair<long long,long long>> cows_raw(N);
    for (auto &c : cows_raw) cin >> c.first >> c.second;

    if (N == 0) {
        cout << 0 << '\n';
        return 0;
    }

    // Compress x
    sort(cows_raw.begin(), cows_raw.end());
    int new_x = 0;
    long long actual_x = cows_raw[0].first;
    for (int i = 0; i < N; ++i) {
        if (cows_raw[i].first != actual_x) {
            ++new_x;
            actual_x = cows_raw[i].first;
        }
        cows_raw[i].first = new_x;
    }

    // Compress y
    sort(cows_raw.begin(), cows_raw.end(),
        [&](const pair<long long,long long>& a, const pair<long long,long long>& b) {
            return pair<long long,long long>(a.second, a.first)
                 < pair<long long,long long>(b.second, b.first);
        });

    int new_y = 0;
    long long actual_y = cows_raw[0].second;
    for (int i = 0; i < N; ++i) {
        if (cows_raw[i].second != actual_y) {
            ++new_y;
            actual_y = cows_raw[i].second;
        }
        cows_raw[i].second = new_y;
    }

    // After compression, all coords <= N, safe to cast to int
    vector<pair<int,int>> cows(N);
    for (int i = 0; i < N; ++i)
        cows[i] = { (int)cows_raw[i].first, (int)cows_raw[i].second };

    // restore order sorted by x,y
    sort(cows.begin(), cows.end());

    int low = 1, high = N + 1;

    while (low < high) {
        int mid = (low + high) / 2;
        bool possible = false;

        auto west_south_height = calc_max_height(mid, cows);

        for (auto &c : cows) c.second = N - 1 - c.second;
        auto west_north_height = calc_max_height(mid, cows);

        reverse(cows.begin(), cows.end());
        for (auto &c : cows) {
            c.first = N - 1 - c.first;
            c.second = N - 1 - c.second;
        }
        auto east_south_height = calc_max_height(mid, cows);

        for (auto &c : cows) c.second = N - 1 - c.second;
        auto east_north_height = calc_max_height(mid, cows);

        for (int x = 0; x < N - 1; ++x) {
            int south = min(west_south_height[x], east_south_height[N - 2 - x]) + 1;
            int north = min(west_north_height[x], east_north_height[N - 2 - x]) + 1;
            if (south + north >= N) {
                possible = true;
                break;
            }
        }

        if (possible) high = mid;
        else low = mid + 1;
    }

    cout << low << '\n';
    return 0;
}

File: contests/problemC/solutions/solution_binary_search.py
import sys
from typing import List


def calc_max_height(max_cows: int, cows: List[List[int]]) -> List[int]:
    n = len(cows)
    count = [0] * n
    bound = [0] * n
    y_bound = n - 1
    total = 0
    i = 0
    for x in range(n):
        while i < n and cows[i][0] == x:
            y = cows[i][1]
            count[y] += 1
            if y <= y_bound:
                total += 1
            i += 1
        while total > max_cows:
            total -= count[y_bound]
            y_bound -= 1
        bound[x] = y_bound
    return bound


def main() -> None:
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    it = iter(data)
    _H = int(next(it))
    _W = int(next(it))
    N = int(next(it))
    cows_raw: List[List[int]] = []
    for _ in range(N):
        cows_raw.append([int(next(it)), int(next(it))])

    if N == 0:
        print(0)
        return

    cows_raw.sort()
    current_x = cows_raw[0][0]
    new_x = 0
    for cow in cows_raw:
        if cow[0] != current_x:
            new_x += 1
            current_x = cow[0]
        cow[0] = new_x

    cows_raw.sort(key=lambda pair: (pair[1], pair[0]))
    current_y = cows_raw[0][1]
    new_y = 0
    for cow in cows_raw:
        if cow[1] != current_y:
            new_y += 1
            current_y = cow[1]
        cow[1] = new_y

    cows = [[int(cow[0]), int(cow[1])] for cow in cows_raw]
    cows.sort()

    low, high = 1, N + 1
    while low < high:
        mid = (low + high) // 2
        possible = False

        west_south_height = calc_max_height(mid, cows)

        for cow in cows:
            cow[1] = N - 1 - cow[1]
        west_north_height = calc_max_height(mid, cows)

        cows.reverse()
        for cow in cows:
            cow[0] = N - 1 - cow[0]
            cow[1] = N - 1 - cow[1]
        east_south_height = calc_max_height(mid, cows)

        for cow in cows:
            cow[1] = N - 1 - cow[1]
        east_north_height = calc_max_height(mid, cows)

        for x in range(0, N - 1):
            south = min(west_south_height[x], east_south_height[N - 2 - x]) + 1
            north = min(west_north_height[x], east_north_height[N - 2 - x]) + 1
            if south + north >= N:
                possible = True
                break

        if possible:
            high = mid
        else:
            low = mid + 1

    print(low)


if __name__ == "__main__":
    main()
